src/java/org/apache/cassandra/index/sai/disk/v1/WideRowAwarePrimaryKeyMap.java

```diff
@@ -0,0 +1,162 @@
+    {
+        private FileHandle primaryKeyTrieFile = null;
+        private final ClusteringComparator clusteringComparator;
+
+        public WideRowAwarePrimaryKeyMapFactory(IndexDescriptor indexDescriptor, SSTableReader sstable)
+        {
+            super(indexDescriptor, sstable);
+
+            try
+            {
+                this.primaryKeyTrieFile = indexDescriptor.createPerSSTableFileHandle(IndexComponent.PRIMARY_KEY_TRIE);
+                this.clusteringComparator = indexDescriptor.clusteringComparator;
+            }
+            catch (Throwable t)
+            {
+                throw Throwables.unchecked(Throwables.close(t, primaryKeyTrieFile));
+            }
+        }
+
+        @Override
+        @SuppressWarnings({ "resource", "RedundantSuppression" })
+        public PrimaryKeyMap newPerSSTablePrimaryKeyMap() throws IOException
+        {
+            LongArray rowIdToToken = new LongArray.DeferredLongArray(tokenReaderFactory::open);
+            return new WideRowAwarePrimaryKeyMap(rowIdToToken,
+                                                 new SortedTermsTrieSearcher(primaryKeyTrieFile.instantiateRebufferer(null), sortedTermsMeta),
+                                                 sortedTermsReader.openCursor(),
+                                                 partitioner,
+                                                 primaryKeyFactory,
+                                                 clusteringComparator);
+        }
+    }
+
+    private final SortedTermsTrieSearcher trieSearcher;
+    private final ClusteringComparator clusteringComparator;
+
+    private WideRowAwarePrimaryKeyMap(LongArray tokenArray,
+                                      SortedTermsTrieSearcher trieSearcher,
+                                      SortedTermsReader.Cursor cursor,
+                                      IPartitioner partitioner,
+                                      PrimaryKey.Factory primaryKeyFactory,
+                                      ClusteringComparator clusteringComparator)
+    {
+        super(tokenArray, cursor, partitioner, primaryKeyFactory);
+
+        this.trieSearcher = trieSearcher;
+        this.clusteringComparator = clusteringComparator;
+    }
+
+    @Override
+    public long rowIdFromPrimaryKey(PrimaryKey key)
+    {
+        long rowId = tokenArray.findTokenRowID(key.token().getLongValue());
+        // If the key only has a token (initial range skip in the query) or the token is out of range
+        // the return the rowId from the token array.
+        if (key.isTokenOnly() || rowId == -1)
+            return rowId;
+        // If the token for the rowId matches the key token then we need to search the keys in the trie
+        // for a prefix match to get the correct rowId for the clustering key.
+        if (tokenArray.get(rowId) == key.token().getLongValue())
+            return trieSearcher.prefixSearch(key::asComparableBytes);
```

Comment:
I'm trying to remember if we actually need to do this for correctness, or it's just an optimization for partitions w/ tons of rows, given this method is just used for skipping. If it's the latter, I wonder how small the partition has to be for accessing the trie not to be worth it...

Reply:
No, this is wrong. I think I got confused when splitting the primary key maps and left this in. If the key isn't token only then we need to full prefix search in all cases.